File: ocaml/AC25.ml
let parse_line line =
  Scanf.sscanf line "%s@: %s@\n" (fun lhs rhs ->
      let rhs_list = String.split_on_char ' ' rhs in
      Common.Atom.get lhs, List.map Common.Atom.get rhs_list)

let add_directed_edge a b graph =
  Common.AtomMap.update a (fun neighbors ->
    Some (Common.AtomSet.add b
      (Option.value ~default:Common.AtomSet.empty neighbors)))
    graph

let add_edge graph (a, b) =
  add_directed_edge a b (add_directed_edge b a graph)

let remove_directed_edge a b graph =
  Common.AtomMap.update a (Option.map (Common.AtomSet.remove b)) graph

let remove_edge (a, b) graph =
  remove_directed_edge a b (remove_directed_edge b a graph)

module EdgeSet = Set.Make (Common.OrderedPair (Common.Atom) (Common.Atom))

let rec bfs_add_neighbors
          flow succ k return u (neighbors : Common.Atom.t Seq.t) pred q =
  match neighbors () with
  | Nil -> k pred q
  | Cons (v, neighbors) ->
     if Common.AtomMap.mem v pred || flow u v then
       bfs_add_neighbors flow succ k return u neighbors pred q
     else
       begin
         let pred = Common.AtomMap.add v u pred in
         if Common.AtomMap.mem v succ then
           return v pred
         else
           bfs_add_neighbors flow succ k return u neighbors pred (v :: q)
       end

let rec bfs_dequeue graph flow succ k return q_s pred q =
  match q_s with
  | [] ->
     if q = [] then
       None
     else
       k pred q
  | u :: q_s ->
     let neighbors = Common.AtomMap.find u graph in
     bfs_add_neighbors flow succ (bfs_dequeue graph flow succ k return q_s)
       return u (Common.AtomSet.to_seq neighbors) pred q

let rec bidirectional_bfs_loop graph flow pred q_s succ q_t =
  if List.length q_s < List.length q_t then
    bfs_dequeue graph (fun u v -> EdgeSet.mem (u, v) flow) succ
      (fun pred q -> bidirectional_bfs_loop graph flow pred q succ q_t)
      (fun v pred -> Some (v, pred, succ))
      q_s pred []
  else
    bfs_dequeue graph (fun u v -> EdgeSet.mem (v, u) flow) pred
      (fun succ q -> bidirectional_bfs_loop graph flow pred q_s succ q)
      (fun v succ -> Some (v, pred, succ))
      q_t succ []

let bidirectional_bfs graph s t flow =
  bidirectional_bfs_loop graph flow (Common.AtomMap.singleton s s) [s]
    (Common.AtomMap.singleton t t) [t]

let[@tail_mod_cons] rec trace_path succ src tgt =
  if src = tgt then
    [tgt]
  else
    src :: trace_path succ (Common.AtomMap.find src succ) tgt

let rec augment flow path =
  match path with
  | [] | [_] -> flow
  | u :: v :: path ->
     augment (EdgeSet.add (u, v) flow) (v :: path)

let rec edmonds_karp_loop graph s t flow value =
  match bidirectional_bfs graph s t flow with
  | None -> value, flow
  | Some (v, pred, succ) ->
     let backward = trace_path pred v s in
     let forward = trace_path succ v t in
     let path = List.rev_append backward forward in
     let flow = augment flow path in
     edmonds_karp_loop graph s t flow (value + 1)
     
let edmonds_karp graph s t =
  edmonds_karp_loop graph s t EdgeSet.empty 0  

let reduce op list =
  match list with
  | [] -> None
  | hd :: tl -> Some (List.fold_left op hd tl)

let min_by cmp list =
  reduce (fun a b -> if cmp a b <= 0 then a else b) list

let min_by_key cmp key list =
  let op a b = cmp (key a) (key b) in 
  min_by op list

let dominating_set graph (node, dominated_nodes) =
  let rec add_remaining_nodes dominating_set remaining_nodes =
    match Common.AtomMap.choose remaining_nodes with
    | exception Not_found -> dominating_set
    | (node, dominated_nodes) ->
       add_remaining_nodes (Common.AtomSet.add node dominating_set)
         (Common.AtomSet.fold Common.AtomMap.remove dominated_nodes
            (Common.AtomMap.remove node remaining_nodes)) in
  add_remaining_nodes (Common.AtomSet.singleton node)
    (Common.AtomSet.fold Common.AtomMap.remove dominated_nodes
       (Common.AtomMap.remove node graph))

let check_dominating_set graph (node, dominated_nodes) =
  let set = dominating_set graph (node, dominated_nodes) in
  if Common.AtomSet.cardinal set >= 2 then
    Some set
  else
    None

let connected_component graph node =
  let rec add_neighbor node accu_set =
    let set = Common.AtomSet.diff (Common.AtomMap.find node graph) accu_set in
    Common.AtomSet.fold add_neighbor set (Common.AtomSet.union accu_set set) in
  add_neighbor node (Common.AtomSet.singleton node)

let nodes_of_graph graph =
  Common.AtomSet.of_seq (Seq.map fst (Common.AtomMap.to_seq graph))

let minimum_st_edge_cut graph s t =
  let value, residual = edmonds_karp graph s t in
  let graph = EdgeSet.fold remove_edge residual graph in
  let reachable = connected_component graph s in
  let non_reachable = Common.AtomSet.diff (nodes_of_graph graph) reachable in
  value, (reachable, non_reachable)

let minimum_edge_cut graph =
  let nodes = Common.AtomMap.bindings graph in
  match List.find_map (check_dominating_set graph) nodes with
  | None ->
     (* This case does not occur in AOC inputs,
        but we handle it for completeness *)
     let node_with_min_degree, _neighbors =
       fst (Option.get (min_by_key Int.compare snd
         (List.map (fun (label, wires) ->
           ((label, wires), Common.AtomSet.cardinal wires)) nodes))) in
     Common.AtomSet.singleton node_with_min_degree,
     Common.AtomSet.remove node_with_min_degree (nodes_of_graph graph)
  | Some set ->
     let v = Common.AtomSet.choose set in
     let cuts =
       List.map (fun w -> minimum_st_edge_cut graph v w)
         (Common.AtomSet.elements (Common.AtomSet.remove v set)) in
     snd (Option.get (min_by_key Int.compare fst cuts))

let part1 wires =
  let wire_pairs =
    List.concat_map (fun (lhs, rhs) -> List.map (fun c -> (lhs, c)) rhs)
      wires in
  let connections =
    List.fold_left add_edge Common.AtomMap.empty wire_pairs in
  let group1, group2 = minimum_edge_cut connections in
  Common.AtomSet.cardinal group1 * Common.AtomSet.cardinal group2

let () =
  let wires = List.of_seq (Seq.map parse_line (Common.input_lines ())) in
  let result_part1 = part1 wires in
  Printf.eprintf "Part 1: %d\n" result_part1

File: rust/src/bin/ac25.rs
use advent_of_code::{Result, NameTable, Name, NameSet, NameMap};

struct Graph(NameMap<NameSet>);

type EdgeSet = std::collections::HashSet<(Name, Name)>;

enum AddNeighbors {
    NextQueue(Vec<Name>),
    Path(Vec<Name>),
}

fn compute_half_path(
    path: &mut Vec<Name>, mut node: Name, succ: &NameMap<Name>
) {
    loop {
        let &next = succ.get(&node).unwrap();
        if next == node {
            break;
        }
        node = next;
        path.push(node);
    }
}

fn compute_path(
    node: Name, pred: &NameMap<Name>, succ: &NameMap<Name>
) -> Vec<Name> {
    let mut result = vec![];
    result.push(node);
    compute_half_path(&mut result, node, pred);
    result.reverse();
    compute_half_path(&mut result, node, succ);
    result
}

impl Graph {
    fn new() -> Self {
        Self(NameMap::new())
    }

    fn add_directed_edge(&mut self, u: Name, v: Name) {
        self.0.entry(u).or_insert_with(NameSet::new).insert(v);
    }

    fn add_edge(&mut self, u: Name, v: Name) {
        self.add_directed_edge(u, v);
        self.add_directed_edge(v, u);
    }

    fn parse_and_add_edges(
        &mut self, table: &mut NameTable, s: &str
    ) -> Result<()> {
        let (src_str, rhs) = s.split_once(": ").ok_or("No ': ' found")?;
        let src = table.get(src_str);
        let tgts_str = rhs.split(" ");
        for tgt_str in tgts_str {
            self.add_edge(src, table.get(tgt_str));
        }
        Ok(())
    }

    fn nodes(&self) -> impl Iterator<Item = Name> + '_ {
        self.0.keys().cloned()
    }

    fn neighbors(&self, node: Name) -> Option<&NameSet> {
        self.0.get(&node)
    }

    fn dominating_set(&self, node: Name) -> NameSet {
        let mut dominating_nodes = NameSet::from([node]);
        let mut dominated_nodes = NameSet::new();
        if let Some(set) = self.neighbors(node) {
            dominated_nodes.extend(set);
        }
        let mut remaining_nodes: NameSet = self.nodes().collect();
        remaining_nodes.remove(&node);
        remaining_nodes =
            remaining_nodes.difference(&dominated_nodes).cloned().collect();
        while let Some(&node) = remaining_nodes.iter().next() {
            remaining_nodes.remove(&node);
            dominating_nodes.insert(node);
            if let Some(neighbors) = self.neighbors(node) {
                dominated_nodes.extend(neighbors);
                remaining_nodes =
                    remaining_nodes.difference(neighbors).cloned().collect();
            }
        }
        dominating_nodes
    }

    fn connected_component(&self, node: Name, residual: &EdgeSet) -> NameSet {
        let mut result = NameSet::from([node]);
        let mut queue = vec![node];
        while let Some(node) = queue.pop() {
            let Some(neighbors) = self.neighbors(node) else { continue };
            let set: NameSet =
                neighbors.difference(&result).cloned()
                .filter(|&tgt|
                    !residual.contains(&(node, tgt)) &&
                    !residual.contains(&(tgt, node))).collect();
            result.extend(&set);
            queue.extend(set);
        }
        result
    }

    fn add_neighbors(
        &self, flow: impl Fn(Name, Name) -> bool, succ: &NameMap<Name>,
        q_s: &mut Vec<Name>, pred: &mut NameMap<Name>
    ) -> AddNeighbors {
        let mut q = vec![];
        while let Some(u) = q_s.pop() {
            let Some(neighbors) = self.neighbors(u) else { continue };
            for &v in neighbors {
                if pred.contains_key(&v) || flow(u, v) {
                    continue;
                }
                pred.insert(v, u);
                if succ.contains_key(&v) {
                    return AddNeighbors::Path(compute_path(v, &pred, succ));
                }
                q.push(v);
            }
        }
        AddNeighbors::NextQueue(q)
    }

    fn bidirectional_bfs(
        &self, s: Name, t: Name, flow: &EdgeSet
    ) -> Option<Vec<Name>> {
        let mut pred = NameMap::from([(s, s)]);
        let mut q_s = vec![s];
        let mut succ = NameMap::from([(t, t)]);
        let mut q_t = vec![t];
        loop {
            if q_s.len() < q_t.len() {
                match self.add_neighbors(
                    |u, v| flow.contains(&(u, v)), &succ, &mut q_s, &mut pred
                ) {
                    AddNeighbors::Path(p) => return Some(p),
                    AddNeighbors::NextQueue(q) => {
                        if q.is_empty() {
                            return None;
                        }
                        q_s = q;
                    }
                }
            }
            else {
                match self.add_neighbors(
                    |u, v| flow.contains(&(v, u)), &pred, &mut q_t, &mut succ
                ) {
                    AddNeighbors::Path(mut p) => {
                        p.reverse();
                        return Some(p)
                    }
                    AddNeighbors::NextQueue(q) => {
                        if q.is_empty() {
                            return None;
                        }
                        q_t = q;
                    }
                }
            }
        }
    }

    fn edmonds_karp(&self, s: Name, t: Name) -> (usize, EdgeSet) {
        let mut flow = EdgeSet::new();
        let mut value = 0;
        while let Some(path) = self.bidirectional_bfs(s, t, &flow) {
            flow.extend(path.iter().cloned().zip(path.iter().skip(1).cloned()));
            value += 1;
        }
        (value, flow)
    }

    fn minimum_edge_cut(&self) -> Result<(NameSet, NameSet)> {
        match self.nodes().find_map(|node|
            Some(self.dominating_set(node)).filter(|set| set.len() >= 2)
        ) {
            None => {
                let (&node_with_min_degree, _neighbors) =
                    self.0.iter()
                        .min_by_key(|(_node, neighbors)| neighbors.len())
                        .ok_or("Empty graph")?;
                let mut other_nodes: NameSet = self.nodes().collect();
                other_nodes.remove(&node_with_min_degree);
                Ok((NameSet::from([node_with_min_degree]), other_nodes))
            }
            Some(mut set) => {
                let &v = set.iter().next().ok_or("Unexpected empty set")?;
                set.remove(&v);
                let (_value, residual) =
                    set.iter().map(|&w| self.edmonds_karp(v, w))
                    .min_by_key(|(value, _residual)| *value)
                    .ok_or("Unexpected empty set")?;
                let reachable = self.connected_component(v, &residual);
                let all_nodes: NameSet = self.nodes().collect();
                let non_reachable: NameSet =
                    all_nodes.difference(&reachable).cloned().collect();
                Ok((reachable, non_reachable))
            }
        }
    }
}


fn main() -> Result<()> {
    let mut table = NameTable::new();
    let mut graph = Graph::new();
    for line in std::io::stdin().lines() {
        graph.parse_and_add_edges(&mut table, &line?)?;
    }
    let (group1, group2) = graph.minimum_edge_cut()?;
    let result_part1 = group1.len() * group2.len();
    println!("Part 1: {result_part1}");
    Ok(())
}
